clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,254 @@
+defConfig :: KnownNat depth => DcConfig (SNat depth)
+defConfig = DcConfig
+  { dcDepth = SNat
+  , dcReadDataCount = True
+  , dcWriteDataCount = True
+  , dcOverflow = False
+  , dcUnderflow = False
+  }
+
+data FifoState n = FifoState
+  { hsQueue      :: Seq.Seq (BitVector n)
+  , relativeTime :: Int -- ^ In order to model a FIFO in two clock domains,
+                        -- we track relative offset in order to know which
+                        -- signals to peel off.
+  } deriving Show
+
+data XilinxFifo read write depth n =
+  XilinxFifo
+    {
+    -- | @full@. Full Flag: When asserted, this signal indicates that the FIFO
+    -- is full. Write requests are ignored when the FIFO is full, initiating a write when
+    -- the FIFO is full is not destructive to the contents of the FIFO.
+      isFull :: Signal write Full
+    -- | @overflow@. Overflow: This signal indicates that a write request
+    -- (wr_en) during the prior clock cycle was rejected, because the FIFO is full.
+    -- Overflowing the FIFO is not destructive to the FIFO.
+    , isOverflow :: Signal write Bool
+    -- | @wr_data_count@. Write Data Count: This bus indicates the number of words
+    -- written into the FIFO. The count is guaranteed to never under-report the number
+    -- of words in the FIFO, to ensure you never overflow the FIFO. The exception to
+    -- this behavior is when a write operation occurs at the rising edge of wr_clk/ clk,
+    -- that write operation will only be reflected on wr_data_count at the next rising clock edge.
+    --
+    -- If D is less than log2(FIFO depth)-1, the bus is truncated by removing the least-significant bits.
+    , writeCount :: Signal write (DataCount depth)
+    -- | @empty@. Empty Flag: When asserted, this signal indicates that the FIFO is empty.
+    -- Read requests are ignored when the FIFO is empty, initiating a read while empty
+    -- is not destructive to the FIFO.
+    , isEmpty :: Signal read Empty
+    -- | @underflow@. Underflow: Indicates that read request (rd_en) during the
+    -- previous clock cycle was rejected because the FIFO is empty. Underflowing the FIFO
+    -- is not destructive to the FIFO.
+    , isUnderflow :: Signal read Bool
+    -- | @rd_data_count@. Read Data Count: This bus indicates the number of words
+    -- available for reading in the FIFO. The count is guaranteed to never over-report the
+    -- number of words available for reading, to ensure that you do not underflow the FIFO.
+    -- The exception to this behavior is when the read operation occurs at the rising edge
+    -- of rd_clk/clk, that read operation is only reflected on rd_data_count at the next
+    -- rising clock edge.
+    --
+    -- If C is less than log2(FIFO depth)-1, the bus is truncated by removing the least-significant bits.
+    , readCount :: Signal read (DataCount depth)
+    -- | @dout@. Data Output: The output data bus is driven when reading the FIFO.
+    , fifoData :: Signal read (BitVector n)
```

Comment:
Can't we just support any `NFDataX a => a` rather than just `BitVector`?

Reply:
We actually do use `n` in `dcFifoBBF`!

I think we'd need something like `sizeof`

Reply:
You could always add something like `KnownNat n, n ~ BitSize a` to the signature, but I think you can get size information in the DSL as well.

Reply:
Ah, thank you! That's more flexible.